arangod/Cluster/Maintenance.cpp

```diff
@@ -476,76 +506,128 @@ VPackBuilder getShardMap(VPackSlice const& plan) {
+  // Sie betreten den funktionalen Sektor.
+
   arangodb::Result result;
   std::unordered_set<std::string> commonShrds;  // Intersection collections plan&local
   std::unordered_set<std::string> indis;  // Intersection indexes plan&local
 
   // Plan to local mismatch ----------------------------------------------------
   // Create or modify if local databases are affected
-  auto pdbs = plan.get(DATABASES);
-  for (auto const& pdb : VPackObjectIterator(pdbs, true)) {
-    if (!local.hasKey(pdb.key.stringRef())) {
-      auto dbname = pdb.key.copyString();
+  for (auto p : plan) {
+    auto const& dbname = p.first;
+    auto pb = p.second->slice()[0];
+    auto const& pdb = pb.get(
+      std::vector<std::string>{AgencyCommHelper::path(), PLAN, DATABASES, dbname});
+
+    if (pdb.isObject() && local.find(dbname) == local.end()) {
       if (errors.databases.find(dbname) == errors.databases.end()) {
-        actions.emplace_back(std::make_shared<ActionDescription>(
-              std::map<std::string, std::string>{{std::string(NAME), std::string(CREATE_DATABASE)}, {std::string("tick"), std::to_string(TRI_NewTickServer())},
-               {std::string(DATABASE), std::move(dbname)}},
-              HIGHER_PRIORITY, false,
-              std::make_shared<VPackBuilder>(pdb.value)));
+        makeDirty.insert(dbname);
+        callNotify = true;
+        actions.emplace_back(
+          std::make_shared<ActionDescription>(
+            std::map<std::string, std::string>{
+              {std::string(NAME), std::string(CREATE_DATABASE)},
+              {std::string(DATABASE), std::move(dbname)}},
+            HIGHER_PRIORITY, false,
+                               std::make_shared<VPackBuilder>(pdb)));
       } else {
         LOG_TOPIC("3a6a8", DEBUG, Logger::MAINTENANCE)
-            << "Previous failure exists for creating database " << dbname << "skipping";
+          << "Previous failure exists for creating database " << dbname << "skipping";
       }
     }
   }
 
-  // Drop databases, which are no longer in plan
-  for (auto const& ldb : VPackObjectIterator(local, true)) {
-    auto dbname = ldb.key.copyString();
-    if (!plan.hasKey(std::vector<std::string>{DATABASES, dbname})) {
-      actions.emplace_back(std::make_shared<ActionDescription>(
-          std::map<std::string, std::string>{{std::string(NAME), std::string(DROP_DATABASE)}, {std::string("tick"), std::to_string(TRI_NewTickServer())},
-           {std::string(DATABASE), std::move(dbname)}}, 
-          HIGHER_PRIORITY, false));
+  // Drop databases, which are no longer in plan ONLY DIRTY
+  for (auto const& dbname : dirty) {
+    if (local.find(dbname) != local.end()) {
+      bool needDrop = false;
+      auto it = plan.find(dbname);
+      if (it == plan.end()) {
+        needDrop = true;
+      } else {
+        VPackSlice pb = it->second->slice()[0];
+        VPackSlice pdb = pb.get(
+            std::vector<std::string>{AgencyCommHelper::path(), PLAN, DATABASES, dbname});
```

Comment:
why to copy `std::string`s, can't it just be a `std::vector<std::string_view>` or `std::vector<arangodb::StringRef>`?

Reply:
not implemented in Slice